### PR TITLE
[appmanifest] add tests for shortcut member

### PR DESCRIPTION
**Pull request by christianliebel**

### CODE REVIEW

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest
Comment (marcoscaceres) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest:2]: We might need to add some additional things to make sure Windows can install this app. 

@aarongustafson can probably advice on what's needed. 

```suggestion
    "name": "Shortcut test",
    "scope": "/", 
    "shortcuts": [{
```

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-invalid-url.webmanifest
Comment (marcoscaceres) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-invalid-url.webmanifest:2]: As above... making sure we don't get a false positive or a false fail because of missing manifest members that Edge likes.

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-out-of-scope-url.webmanifest
Comment (marcoscaceres) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-out-of-scope-url.webmanifest:2]: I'd suggest adding an explicit scope everywhere. 

```suggestion
    "scope": "/",
    "shortcuts": [{
```

**Review by marcoscaceres — CHANGES_REQUESTED** [review]
Generally looking good. Definitely want @aarongustafson to check these tho.

**Review by marcoscaceres — COMMENTED** [review]
Wondering... is there some way for us to test the short_name or does name always win? I think you mentioned this yesterday... looking now, it is weird that name is require if you can just have a short_name... it should maybe be that one or the other should be there.

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest
Comment (aarongustafson) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest:14]: Empty URL means whatever the scope is, right (based on shortcut URL calculation)?

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-out-of-scope-url.webmanifest
Comment (aarongustafson) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-out-of-scope-url.webmanifest:5]: Should the name here be "fail"?

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-invalid-url.webmanifest
Comment (aarongustafson) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-invalid-url.webmanifest:5]: Should this be named "fail"?

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-undefined-name.webmanifest
Comment (aarongustafson) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-undefined-name.webmanifest:5]: short_name: fail?

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-undefined-url.webmanifest
Comment (aarongustafson) [inline appmanifest/shortcuts-member/shortcuts-member-skip-for-undefined-url.webmanifest:5]: name: fail

**Review by aarongustafson — COMMENTED** [review]
This is looking great. Just a few questions from my side.

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest
Comment (christianliebel) [reply appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest:14]: > Set shortcut["url"] to the result of parsing shortcut["url"] using manifest URL as the base URL. If the result is failure, issue a developer warning and continue.

From my understanding, an empty URL would link to the manifest file (see [processing the shortcuts member, step 3](https://www.w3.org/TR/appmanifest/#dfn-processing-the-shortcuts-member)). Not sure if that's cool though. We might want to add a test for this special case.

**Comment by christianliebel** [comment]
> Wondering... is there some way for us to test the short_name or does name always win?

I guess we can't really test `short_name` unless we define a certain environment that only shows `short_name`s. 

> I think you mentioned this yesterday... looking now, it is weird that name is require if you can just have a short_name... it should maybe be that one or the other should be there.

Yes, I think what I said there was that the shortcuts' `name` and `short_name` properties work a little differently than the same properties of the manifest. Shortcuts require a `name` which can be overridden by `short_name` based on the environment (https://www.w3.org/TR/appmanifest/#shortcuts-member), in the manifest where you can specify one of which or both (https://www.w3.org/TR/appmanifest/#application-s-name). That could be a bit counter-intuitive.

.cc @aarongustafson

**Comment by aarongustafson** [comment]
> Yes, I think what I said there was that the shortcuts' name and short_name properties work a little differently than the same properties of the manifest. 

I *think* the manifest’s lack of a required `name` stems from it’s lack of requirement for **any** member. Is that right @marcoscaceres? If I were starting fresh with the manifest spec, I think I’d have made `name` required, which is why I went that route for `ShortcutItem`. Happy to hear counterarguments to that though. The main thing is that a shortcut can’t exist without a label.

File: appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest
Comment (marcoscaceres) [reply appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest:14]: That's developer error, I think. It's like doing the following in CSS:

```
body {
  background-image: url(""); 
}
```

Comment (aarongustafson) [reply appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest:14]: As such, should we suggest UAs provide a warning in the console?

Comment (marcoscaceres) [reply appmanifest/shortcuts-member/shortcuts-member-skip-for-empty-name.webmanifest:14]: No, as it's not really an error. However, a UA would be free to show something in dev tools, but it wouldn't be something we would normally put in the spec.

File: appmanifest/shortcuts-member/shortcuts-member.webmanifest
Comment (marcoscaceres) [inline appmanifest/shortcuts-member/shortcuts-member.webmanifest:8]: I think we should have a "shortcut_pass.html" file instead. We could then use that throughout the tests to make sure the shortcut is landing in the right place. 

The file could just be:

```HTML 
<html>
<meta charset="utf-8">
<body style="background: green">
<h1>Pass</h1>
```

Similarly, we could have a fail version of the above with a red background.

**Review by marcoscaceres — APPROVED** [review]

**Comment by marcoscaceres** [comment]
closing and reopening to see if it restarts CI, which seems to be stuck :/

**Comment by marcoscaceres** [comment]
@christianliebel, sorry, can you rebase... that might get it going again. Then we should be good to merge.

**Comment by marcoscaceres** [comment]
Actually, CI restarted... we might be good... 🤞

**Comment by stephenmcgruer** [comment]
I'm not sure what Taskcluster is up to here (it doesn't seem to have reported one of the required tasks to GitHub), but the rest of the subtasks look good so I'm happy to admin merge this if/when ready - just let me know when I should click the merge button :)

**Comment by marcoscaceres** [comment]
@stephenmcgruer appreciate that. Please go ahead and merge.

**Pull request merged**
